submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,354 @@
+from the backend Pods will have the source as the globalIP of the exportedService.
+
+On the other hand, when the backend Pod tries to talk to the external application in remote
+Cluster, the external application will see that traffic is coming with a globalIP that is
+assigned either at the Cluster level, or namespace level, or the one requested by the user
+at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them and should
+avoid requesting explicit globalIP that match Headless Service backend Pods.
+
+The proposal has its own set of Pros and Cons.
+
+### Pros
+
+Since we will be annotating only select K8s objects with globalIP, it will help in improved scalability.
+Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
+improve the overall performance of the solution.
+
+### Cons
+
+A limitation with the proposed approach is that we will lose the unique identity for **every**
+Pod which is supported today. For users that intend to use Network Policies across Clusters or
+for debugging purposes, this can be mitigated by creating the necessary GlobalnetEgressIP CRDs
+which create unique globalIPs at the Namespace/Pod level.
+
+### Caveats
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type GlobalnetEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
+    Spec               EgressIPSpec `json:"spec"`
+
+    // Observed status of GlobalnetEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type EgressIPSpec struct {
+    // User can explicitly request a specific GlobalIP and this should be a valid IPaddress
+    // from globalnetCIDR allocated to the Cluster. If the requested globalIP falls outside
+    // of the globalnetCIDR or is already allocated (either fully or partially), then Status
+    // field will be set to Error with appropriate message.
+    // If this field is not specified, a single globalIP will be allocated.
+    // This field cannot be changed through updates.
+    // +optional
+    GlobalIP      string  `json:"globalIP,omitempty"`
```

Comment:
After thinking about it some more, how much sense does it really make to allow to specify an explicit IP address here?

In K8s AFAIK, except very specific cases of using LoadBalancer Services on specific clouds, you can't specify an IP address. K8s takes care of IP address allocations.

In ingress global IP use case, the globalnet takes care of IP allocation.

Wouldn't it make more sense for the system to allocate the egress IP as well?

For firewall considerations the users can retrieve the IP once allocated and use that to define the necessary firewall rules.

For a system perspective, allowing to specify the IP seems like a more risky endeavour than having globalnet control allocation (e.g. it can more effectively prevent race conditions, for example).

Reply:
Specifying GlobalIPs provides ability to:
1. Keep the GlobalIP unchanged,
2. Specify the same GlobalIP to certain resources.

If these abilities are provided by the other way, directly specifying the globalIPs might not be necessary (I'm not sure that I created an exclusive list, so someone might find other requirements, later).

Note that I assume that users would expect the globalIP to be unchanged beyond the lifecycle of the namespace or resources inside it. And it reminds me of abstractions like SC/PVC/PV and Elastic IP, which might be too complex for users in this case.

Reply:
> It seems though that he's fine with not having the ability to specify a specific IP, as long as the IP stays consistent (which it should).

Indeed, the globalIP assigned to a Pod/Service will remain consistent through the life-cycle of the resource. @mkimuram can you please confirm if it's okay to remove the ability to specify globalIP as part of the `GlobalnetEgressIP` request?


> > For example, let's say user deployed a Pod and it is assigned some globalIP. User then configures the external-firewall/application/network-policies to use the globalIP. If the Pod gets restarted, unless there is a provision to specify GlobalIP, a different globalIP could be assigned to the Pod.
> 
> Why would that happen? I merely suggested that globalnet controller be allocating the IPs, without allowing the user to do so.

Yeah, when the user creates a `GlobalnetEgressIP` object, a globalIP will be allocated to the matching Pods. As long as the object is not deleted, the same globalIP will be re-used even when the Pods are restarted. 

> We could always add the ability to specify a specific IP in a later stage if thats really a valuable feature, but to me it seems unnecessary for now.

I'm fine with this, if its Okay to others. @mangelajo @mkimuram @aswinsuryan @vthapar

Reply:
@sridhargaddam 

> Indeed, the globalIP assigned to a Pod/Service will remain consistent through the 
> life- cycle of the resource. @mkimuram can you please confirm if it's okay to remove 
> the ability to specify globalIP as part of the GlobalnetEgressIP request?

As commented in the meeting, I agree with it for initial implantation. Let's consider adding it later, if we actually get feedback from users that it is really needed.